base/compiler/typeutils.jl

```diff
@@ -222,3 +222,21 @@ function improvable_via_constant_propagation(@nospecialize(t))
     end
     return false
 end
+
+# convert a Union of Tuple types to a Tuple of Unions
+function unswitchtupleunion(u::Union)
+    ts = uniontypes(u)
+    n = -1
+    for t in ts
+        if t isa DataType && t.name === Tuple.name && !isvarargtype(t.parameters[end])
+            if n == -1
+                n = length(t.parameters)
+            elseif n != length(t.parameters)
+                return u
```

Comment:
we could also bin them by length/type, but I suppose the observation is that this caller doesn't need that?